src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs

```diff
@@ -6231,37 +6384,423 @@ public void Nullable_04()
+
+Program.M().Report();
+
+[CollectionBuilder(typeof(MyCollectionBuilder), nameof(MyCollectionBuilder.Create))]
+public struct MyCollection<T> : IEnumerable<T>
+{
+    private readonly List<T> _list;
+    public MyCollection(List<T> list) { _list = list; }
+    public IEnumerator<T> GetEnumerator() => _list.GetEnumerator();
+    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+}
+
+public class MyCollectionBuilder
+{
+    public static MyCollection<T> Create<T>(ReadOnlySpan<T> items)
+    {
+        return new MyCollection<T>(new List<T>(items.ToArray()));
+    }
+}
+
+partial class Program
+{
+    static MyCollection<int>? M()
+    {
+        return (MyCollection<int>?)[1, 2, 3];
+    }
+}
+""";
+            var comp = CreateCompilation(new[] { src, s_collectionExtensions }, targetFramework: TargetFramework.Net80);
+            comp.VerifyDiagnostics();
+
+            var verifier = CompileAndVerify(comp, expectedOutput: IncludeExpectedOutput("[1, 2, 3],"), verify: Verification.FailsPEVerify);
+            verifier.VerifyIL("Program.M", """
+{
+  // Code size       21 (0x15)
+  .maxstack  1
+  IL_0000:  ldtoken    "<PrivateImplementationDetails>.__StaticArrayInitTypeSize=12_Align=4 <PrivateImplementationDetails>.4636993D3E1DA4E9D6B8F87B79E8F7C6D018580D52661950EABC3845C5897A4D4"
+  IL_0005:  call       "System.ReadOnlySpan<int> System.Runtime.CompilerServices.RuntimeHelpers.CreateSpan<int>(System.RuntimeFieldHandle)"
+  IL_000a:  call       "MyCollection<int> MyCollectionBuilder.Create<int>(System.ReadOnlySpan<int>)"
+  IL_000f:  newobj     "MyCollection<int>?..ctor(MyCollection<int>)"
+  IL_0014:  ret
+}
+""");
+            var tree = comp.SyntaxTrees.First();
+            var model = comp.GetSemanticModel(tree);
+
+            var cast = tree.GetRoot().DescendantNodes().OfType<ReturnStatementSyntax>().Last().Expression;
+            Assert.Equal("(MyCollection<int>?)[1, 2, 3]", cast.ToFullString());
+            var castConversion = model.GetConversion(cast);
+            Assert.True(castConversion.IsIdentity);
+
+            var value = tree.GetRoot().DescendantNodes().OfType<CastExpressionSyntax>().Last().Expression;
+            Assert.Equal("[1, 2, 3]", value.ToFullString());
+            var conversion = model.GetConversion(value);
+            Assert.True(conversion.IsValid);
+            Assert.True(conversion.IsNullable);
+            Assert.False(conversion.IsCollectionExpression);
+
+            Assert.Equal(1, conversion.UnderlyingConversions.Length);
+            var underlyingConversion = conversion.UnderlyingConversions[0];
+            Assert.True(underlyingConversion.IsValid);
+            Assert.False(underlyingConversion.IsNullable);
+            Assert.True(underlyingConversion.IsCollectionExpression);
+
+            var typeInfo = model.GetTypeInfo(value);
+            Assert.Null(typeInfo.Type);
+            Assert.Equal("MyCollection<System.Int32>", typeInfo.ConvertedType.ToTestDisplayString());
```

Comment:
Tagging @AlekseyTs to confirm this is indeed the behavior we want for `ConvertedType` in cast scenario.

Reply:
This is not what I would expect. I would expect the type to be `null`. There is an explicit nullable conversion on top of this, with an underlying implicit collection conversion on the input to it.

Reply:
I see. Let's chat with Aleksey on Monday to confirm. If that's the case, then the previous behavior (absent nullable scenarios) was incorrect.

Reply:
Confirmed the behavior of `ConvertedType` with some other typeless expression. I'll go with that.
```
        [Fact]
        public void TODO2()
        {
            string src = """
class Program
{
    static int? M()
    {
        return (int?)null;
    }
}
""";
            var comp = CreateCompilation(src);
            comp.VerifyDiagnostics();
            var tree = comp.SyntaxTrees.First();
            var model = comp.GetSemanticModel(tree);
            var value = tree.GetRoot().DescendantNodes().OfType<CastExpressionSyntax>().Last().Expression;
            var typeInfo = model.GetTypeInfo(value);
            Assert.Null(typeInfo.Type);
            Assert.Null(typeInfo.ConvertedType);
        }
```